src/BoundsInference.cpp

```diff
@@ -109,27 +110,120 @@ class BoundsInference : public IRMutator {
     set<string> in_pipeline, inner_productions;
     Scope<int> in_stages;
 
+    struct CondValue {
+        Expr cond; // Condition on params only (can't depend on loop variable)
+        Expr expr;
```

Comment:
I can't tell for sure, but the few uses of CondValue I looked at, the choice of Stmt or Expr could be compile time, i.e. CondValue could be a template `CondValue<Stmt>` or `CondValue<Expr>`. I would personally prefer this if possible, it eliminates the possibility of the wrong class being used in one place or another.